LibCrypto+Tests: Fix crashes on invalid datetime strings

BenWiederhake opened the pull request:
Our ASN1 datetime parser used to crash when the trailing 'Z' was missing. Specifically, the code called `GenericLexer::consume()`, without making sure there is something left to consume.

In theory, this means that visiting a malicious website with Browser (which will attempt to validate a TLS certificate, thus parse its ASN1 content, stumble upon the malicious datetime, thus asserting) could crash Browser.

OSS Fuzz found this bug originally in `parse_utc_time`:
- https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=42354

These two crashes are probably the same, since I cannot reproduce these crashes with the fix:
- https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=39139
- https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=46783 (looks different though?)

The same bug was also in `parse_generalized_time` which (hopefully) is the more common one.

This PR also adds tests for these functions, so that they don't regress in the future.

We still make parsing mistakes (ignoring the timezone offset and dropping the milliseconds), but they are much more harmless than a crash.

GMTA commented on Tests/LibCrypto/TestASN1.cpp:
```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright (c) 2021, Ben Wiederhake <BenWiederhake.GitHub@gmx.de>
```
You might want to get with the times ;-)

GMTA commented on Tests/LibCrypto/TestASN1.cpp:
```diff
@@ -0,0 +1,216 @@
+
+private:
+    Core::DateTime m_datetime;
+};
```
What does `DebuggableDateTime` add here? If it's the `operator==`, wouldn't that be a useful addition to `Core::DateTime` instead?

BenWiederhake replied:
Also the `AK::Formatter` for it, but I'll just push both into Datetime.h.

alimpfard commented on Userland/Libraries/LibCrypto/ASN1/ASN1.cpp:
```diff
@@ -84,26 +84,25 @@ Optional<Core::DateTime> parse_utc_time(StringView time)
-            }
+
+    if (lexer.next_is(is_any_of("0123456789"sv))) {
+        seconds = lexer.consume(2).to_uint();
```
This looks like an unchecked assumption about the length again, `yyMMDDHHmms`
Or does consume(n) consume at most n bytes?

BenWiederhake replied:
It's "at most n bytes". I am aware that this means for the `generalized` variant that something like `yyyyMMDDH` (i.e. only a single hour digit) will be accepted as valid, but we don't really worry about that, here.

alimpfard replied:
sure, sgtm in that case.

BenWiederhake commented:
Changes since last push:
- Rebased for convenience
- Fixed the copyright year – d'oh!
- Moved `::operator==` and `AK::Formatter<>` to `Core::DateTime`, since this should be useful also in general

BenWiederhake commented:
Changes since last push:
- Just rebased.

This takes advantage of the more stable CI pipeline due to 76f524c1d60d0ab0ed0da540c4f73505117f8131, and should make this PR green.

BenWiederhake commented:
Changes since last push:
- Just rebased. The old commits still apply cleanly.

Note that this *still* resolves bug https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=42354

This PR was already reviewed, only someone clicking "merge" is missing.
